Updates & additions to reftest documentation

rhauck opened the pull request:
Reviewed and made a few changes to the original version of the reftest doc migrated from the CSSWG wiki: 
- Changed the example reftest to be one that uses red/green
- Migrated the test format guidelines and self describing test docs
- Fixed a bunch of formatting issues
- Added the mention of reftests being self-describing tests back - they can be both, but changed the language in the selftest.md doc to reflect that they are sometimes automatable if they're reftests.
- Removed the mention of 'unreftestable tests' from this doc - IMO, this is not the right place for this info - better trackable through Shepherd. 

Note: I had to use plain HTML instead of MD for the Requirements Flags tabe in the test format guidelines doc - it _would not render_ when built with Jekyll.  Ditto for the anchor links from the TOC in that doc.

I think we should discuss how much of this is CSS-specific info and how to present that info here (so as not to put extra requirements for tests for other specs)

rhauck committed:
test format starter page

rhauck committed:
added TOC

rhauck committed:
filled in some sections

rhauck committed:
completed migrating the format guidelines

rhauck committed:
reviewed and updated reftest doc

rhauck committed:
temporarily removing my forked CNAME file

rhauck committed:
reviewed and updated reftest doc

rhauck committed:
removed backtick codeblock syntax in favor of indentation to make jerky Jekyll happy

rhauck committed:
putting CNAME back

rhauck committed:
trailing newline for parity

rhauck committed:
merging upstream changes

rhauck committed:
fixed code blocks, requirement flags table, anchor links + misc other changes

rhauck committed:
added line wraps for easier reviewability

tobie commented on docs/reftests.md:
````diff
@@ -1,159 +1,119 @@
-`transform` property.
-
-``` html
-<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Strict//EN"
````
I actually liked seeing the code directly in the page rather than having to navigate, here.

rhauck replied:
Fixed in 7af6bd8.

tobie commented on docs/reftests.md:
```diff
@@ -223,41 +182,40 @@ Here is an example of a reftest manifest.
-[7]: mailto:public-css-testsuite@w3.org
+[2]: http://test.csswg.org/source/contributors/adobe/submitted/svg-transform/translate/svg-translate-001.html
+[3]: http://test.csswg.org/source/contributors/adobe/submitted/svg-transform/translate/reference/svg-translate-ref.html
+[4]: ./test-format-guidelines.html
```
I strongly suggest using more descriptive indexes here as these bare change to the doc much better than numerals, e.g. change:

```
[4]: ./test-format-guidelines.html
```

to:

```
[guidelines]: ./test-format-guidelines.html
```

etc.

rhauck replied:
Fixed in 7af6bd8.

tobie commented on docs/selftest.md:
```diff
@@ -0,0 +1,59 @@
+
+# Self Describing Tests
+
+A self-describing test is a test page that describes what the page 
```
I'm not sure why we explain had to build tests we're trying to get rid of.

Should we remove this altogether, or add a big disclaimer?

Should we explain how to convert them to ref tests or scripted tests?

tobie commented on docs/selftest.md:
```diff
@@ -0,0 +1,59 @@
+
+Self-describing tests must follow the [test format guidelines][10].
+
+Additional information on writing self-describing tests is available 
```
We're trying to centralize the doc, here, so let's either include this ref or get rid of it.

tobie commented on docs/test-format-guidelines.md:
```diff
@@ -0,0 +1,879 @@
+layout: default
+title: Test Format
+---
+**Table of Contents** 
```
I'm trying to find a way to automate per page ToC. Will ping you if I do.

tobie commented on docs/test-format-guidelines.md:
```diff
@@ -0,0 +1,879 @@
+<a name="acceptable-test-formats">
+## Acceptable Test Formats
+
+The preferred submission format for CSSWG tests is either XHTML or 
```
Can we make this less CSS WG specific?

rhauck replied:
Fixed in 2659095.

tobie commented on docs/test-format-guidelines.md:
```diff
@@ -0,0 +1,879 @@
+processed by an HTML5 compatible parser. SVG is also acceptable 
+where necessary.
+
+Note that in general, the test source will be parsed and re-
```
Reftest specific.

rhauck replied:
Fixed in 7af6bd8.

tobie commented on docs/test-format-guidelines.md:
```diff
@@ -0,0 +1,879 @@
+Images must be in either PNG or SVG format. (PNG is preferred where 
+raster images can be used.)
+
+A set of scripts will generate the various formats (XHTML, HTML, 
```
CSS WG-specific

rhauck replied:
I'm actually not sure what to do with this bit of information or if it's even worth mentioning at all?  I don't see the harm of removing it. I'll check with @plinss on this one.

plinss replied:
This is specific to the CSSWG at the moment. Once our build code is re-factored and made more generic, I expect other WGs may start using it too. It may be worth keeping the info but changing the 'in general' to something like 'some test suites will be run through a build process that may...'

tobie replied:
I like @plinss's suggestion and wording.

rhauck replied:
fixed in be80612

tobie commented on docs/test-format-guidelines.md:
```diff
@@ -0,0 +1,879 @@
+character encoding must be specified properly per the specification 
+of the source format.
+
+__If the test uses the Ahem font, make sure its computed font-size 
```
Link to Ahem font or explain what it does?

rhauck replied:
Fixed in 7af6bd8. All Ahem info & links are now in test-style-guidelines.md

tobie commented on docs/test-format-guidelines.md:
````diff
@@ -0,0 +1,879 @@
+```
+
+<a name="reviewer">
+### Reviewer
````
CSS WG specific too. Git is responsible for keeping reviewer info.

rhauck replied:
This is another thing we need to figure out. This is more vital info in the CSSWG world.  The question is, should css-specfic stuff all be pulled into css doc or should it just be tagged (with an icon or something) inline?  I think I'm in favor of the latter.

tobie commented on docs/test-format-guidelines.md:
````diff
@@ -0,0 +1,879 @@
+```
+
+<a name="reference-links">
+### Reference Links
````
This probably belongs in the reftest page?

tobie commented on docs/test-format-guidelines.md:
```diff
@@ -0,0 +1,879 @@
+required.
+
+<a name="http-headers">
+## HTTP headers
```
We're [planning](http://lists.w3.org/Archives/Public/public-test-infra/2013JulSep/0010.html) to avoid having to support .htaccess files. Do you have a lot of those in the CSS WG?

rhauck replied:
I haven't yet caught up to the resolution of the thread you started on and in response to this. Frankly, I don't know much about it. Mind if you or someone with more knowledge clarifies this here?  Maybe @plinss?

plinss replied:
I don't think we came to a resolution on this one. There were several suggestions but I didn't see any conclusion.

For the record, we do have a handful of .htaccess files in the CSSWG several of them cover multiple tests. There are about 50 tests currently flagged as needing the headers, I'm not sure how accurately the flags are related the the .htaccess files though.

tobie replied:
Yeah, sorry I should have wrapped this thread up. We're going for a .headers like format. Details of which are still TBD, but very close to Mozilla's description.

tobie commented:
Rough first pass. Overall I feel like `docs/test-format-guidelines.md` has great content, but might need to be broken up into smaller files.

rhauck committed:
A whole lotta changes - fuller description to follow with PR

rhauck commented:
Summary of changes in 7af6bd8:

Bunches and bunches of refactoring and reorganizing, including:

```
1. Changed to reftests.md:
- Re-ordered Reftest doc - components described first, then the example
- Removed reference to self describing tests as a type of test. Instead,
  it's a characteristic of a test this required for rare case manual tests
  and strongly encouraged in reftests.

2. Changed to test-format-guidelines.md & added test-templates.md
  - Pulled test templates + template details the new test-templates
    page  (also added some testharness.js templates for completeness)
    This now makes test-format-guidelines cover just format & minimum 
    general requirements. Note: Git is seeing the new file as a rename of 
    the test-format-guidelines doc - not sure why.

3. Added test-style-guidelines:
- Ported Section 4, "Writing Ideal tests" from
  http://www.w3.org/Style/CSS/Test/guidelines.html
  into this new doc. The other sections of the w3.org doc are already
  included in other pages here. 

4. Removed selftests.md 
   - Added info there to the test-style-guidelines.md and clarified that
   this is required for rare case manual tests, highly preferred for 
   reftests, and OK but not common for script tests.  

5.Added sources.md to track original sources of content - still a bunch of 
   TODO's in here, should be finalized before all this goes live.

6. General changes worth noting:
    - De-emphasized manual testing
    - Subtle use of normative language (particularly re: self describing
    tests)
    - Links to examples all point to files in repo (few exceptions - 
    some still use Hixie's, but no more should point to tests hosted on
    w3.org)
```

rhauck committed:
minor change requested by Tobie that I missed in my previous mondo push

tobie commented on docs/reftests.md:
```diff
@@ -1,159 +1,111 @@
+A _reftest_ is a test that compares the visual output of one file 
+(the test case) with the output of one or more other files (the 
+references). Reftests can be scripted to run and report results 
+automatically, for example, using Web Driver. 
```
spelling: WebDriver.

rhauck replied:
fixed in be80612

rhauck committed:
fixed WebDriver spelling and change language re: building test suites

rhauck commented:
closes w3c/testtwf-website#18
